File: packages/runar-rb/lib/runar/analyzer/sig_analyzer.rb
# frozen_string_literal: true

require_relative 'types'
require_relative 'path_analyzer'

module Runar
  module Analyzer
    # Signature hygiene (spec §9).
    module SigAnalyzer
      def self.analyze(ops, paths)
        findings = []

        # NO_SIG_CHECK: one per reachable path with hasCheckSig == false.
        paths.each do |path|
          next unless path[:reachable]
          next if path[:has_check_sig]
          findings << {
            severity: 'warning',
            code: 'NO_SIG_CHECK',
            message: 'Execution path has no signature verification (OP_CHECKSIG/OP_CHECKMULTISIG)',
            path: path[:description]
          }
        end

        # CHECKSIG_RESULT_DROPPED.
        ops.each_with_index do |op, i|
          next unless op[:opcode] == 0xac || op[:opcode] == 0xae
          next_op = ops[i + 1]
          next unless next_op && next_op[:opcode] == 0x75
          findings << {
            severity: 'warning',
            code: 'CHECKSIG_RESULT_DROPPED',
            message: "#{op[:name]} result is dropped by #{next_op[:name]} — signature check has no effect",
            offset: op[:offset],
            opcode: op[:name]
          }
        end

        findings
      end
    end
  end
end

File: packages/runar-rs/src/analyzer/sig_analyzer.rs
//! Pass 5: signature hygiene. See spec §9.

use super::types::{ExecutionPath, Finding, ParsedOp, Severity};

/// Returns NO_SIG_CHECK and CHECKSIG_RESULT_DROPPED findings.
pub fn analyze_sig_hygiene(opcodes: &[ParsedOp], paths: &[ExecutionPath]) -> Vec<Finding> {
    let mut out: Vec<Finding> = Vec::new();

    // NO_SIG_CHECK per reachable path.
    for path in paths {
        if path.reachable && !path.has_check_sig {
            out.push(Finding {
                severity: Severity::Warning,
                code: "NO_SIG_CHECK".to_string(),
                message: "Execution path has no signature verification (OP_CHECKSIG/OP_CHECKMULTISIG)".to_string(),
                offset: None,
                opcode: None,
                path: Some(path.description.clone()),
            });
        }
    }

    // CHECKSIG_RESULT_DROPPED — only OP_CHECKSIG (0xac) and OP_CHECKMULTISIG
    // (0xae); the *VERIFY variants don't leave a result.
    for i in 0..opcodes.len().saturating_sub(1) {
        let here = &opcodes[i];
        let next = &opcodes[i + 1];
        if (here.opcode == 0xac || here.opcode == 0xae) && next.opcode == 0x75 {
            out.push(Finding {
                severity: Severity::Warning,
                code: "CHECKSIG_RESULT_DROPPED".to_string(),
                message: format!(
                    "{} result is dropped by {} — signature check has no effect",
                    here.name, next.name
                ),
                offset: Some(here.offset),
                opcode: Some(here.name.clone()),
                path: None,
            });
        }
    }

    out
}
